tools/buildgen/extract_metadata_from_bazel_xml.py

```diff
@@ -628,6 +699,52 @@ def _generate_build_extra_metadata_for_tests(
     return test_metadata
 
 
+def _parse_http_archives(xml_tree: ET.Element) -> List[ExternalProtoLibrary]:
+    """Parse Bazel http_archive rule into ExternalProtoLibrary objects."""
+    result = []
+    for xml_http_archive in xml_tree:
+        if xml_http_archive.tag != 'rule' or xml_http_archive.attrib[
+                'class'] != 'http_archive':
+            continue
+        # A distilled Python representation of Bazel http_archive
+        http_archive = dict()
+        for xml_node in xml_http_archive:
+            if xml_node.attrib['name'] == 'name':
+                http_archive["name"] = xml_node.attrib['value']
+            if xml_node.attrib['name'] == 'urls':
+                http_archive["urls"] = []
+                for url_node in xml_node:
+                    http_archive["urls"].append(url_node.attrib['value'])
+            if xml_node.attrib['name'] == 'url':
+                http_archive["urls"] = [xml_node.attrib['value']]
+            if xml_node.attrib['name'] == 'sha256':
+                http_archive["hash"] = xml_node.attrib['value']
+            if xml_node.attrib['name'] == 'strip_prefix':
+                http_archive["strip_prefix"] = xml_node.attrib['value']
+        if http_archive["name"] not in EXTERNAL_PROTO_LIBRARIES:
+            # If this http archive is not one of the external proto libraries,
+            # we don't want to include it as a CMake target
+            continue
+        for url in http_archive["urls"]:
```

Comment:
strictly speaking, duplicating the external_library entries when there's multiple URLs is not correct, but I can understand that it makes the logic in the CMakeLists.txt simpler to implement (but it also makes it a bit harder to understand).

Since the sha256 and strip_prefix for the mirrored urls always needs to be the same, I think it would be better for the `external_library` entity in the build_autogenerated.yaml to allow URL to be a list of URLs (this is not a problem at all for the yaml file). That way, the "metadata" in build_autogenerated.yaml would be more correct (because you'd keep the clear mapping between http_achives and "external_library" entries). The duplication download_archive rules can then be trivially done in CMakeLists.txt.template (and at that point it would be also easier to understand the logic behind it - you could even add an explanatory comment for cases where "urls" contains multiple entries.